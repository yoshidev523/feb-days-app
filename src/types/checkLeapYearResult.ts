export type CheckLeapYearResult =
  | { status: "ok"; isLeapYear: boolean }
  | { status: "error" };
