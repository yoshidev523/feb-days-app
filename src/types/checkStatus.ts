export type CheckStatus = "idle" | "checking" | "error";
