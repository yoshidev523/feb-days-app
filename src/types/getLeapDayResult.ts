export type GetLeapDayResult =
  | { status: "ok", value: number }
  | {status: "error"}