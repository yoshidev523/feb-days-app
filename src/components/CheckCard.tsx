"use client";
import React, { ChangeEvent, useState } from "react";
import { CheckStatus } from "@/types/checkStatus";
import { CheckLeapYearResult } from "@/types/checkLeapYearResult";

export const CheckCard = () => {
  const [message, setMessage] = useState(":)");
  const [checkStatus, setCheckStatus] = useState<CheckStatus>("idle");
  const [inputValue, setInputValue] = useState("");

  const sleep = (msec: number) =>
    new Promise((resolve) => setTimeout(resolve, msec));

  const isLeapYear = (year: number): CheckLeapYearResult => {
    if (year < 0) return { status: "error" };
    if (!Number.isInteger(year)) return { status: "error" };
    return {
      status: "ok",
      isLeapYear: (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0,
    };
  };

  const handleClickCheckButton = async () => {
    setCheckStatus("checking");

    await sleep(2000);

    const inputValueNum = Number(inputValue);
    if (isNaN(inputValueNum)) {
      setCheckStatus("error");
      return;
    }
    const result = isLeapYear(inputValueNum);
    if (result.status === "error") {
      setCheckStatus("error");
      return;
    }
    setCheckStatus("idle");
    setMessage(
      `February ${inputValueNum} has ${result.isLeapYear ? 29 : 28} days.`,
    );
  };

  const handleChangeInput = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  return (
    <div className={"space-y-6 rounded-2xl bg-white p-12"}>
      <div className={"flex justify-between space-x-3"}>
        <input
          type="number"
          onChange={handleChangeInput}
          className={
            "border-b-2 border-gray-800 p-2 focus:border-b-cyan-600 focus:outline-none"
          }
        />
        <button
          className={
            "rounded bg-emerald-500 px-4 py-2 text-white hover:bg-emerald-800"
          }
          onClick={handleClickCheckButton}
        >
          Check
        </button>
      </div>
      <div>
        {checkStatus === "idle" && (
          <p
            className={
              "animate-fade-up text-center text-2xl font-bold"
            }
          >
            {message}
          </p>
        )}
        {checkStatus === "checking" && (
          <p
            className={
              "animate-infinite animate-ease-in-out animate-bounce text-center text-2xl font-bold"
            }
          >
            Checking...
          </p>
        )}
        {checkStatus === "error" && (
          <p
            className={
              "animate-fade-up text-center text-2xl font-bold text-red-600"
            }
          >
            Something Wrong... :(
          </p>
        )}
      </div>
    </div>
  );
};
