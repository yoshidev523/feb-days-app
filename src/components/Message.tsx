import React from "react";
import { CheckStatus } from "@/types/checkStatus";

type MessageProps = {
  checkStatus: CheckStatus;
  message: string;
};

export const Message = ({ checkStatus, message }: MessageProps) => {
  return (
    <div data-testid={"message"}>
      {checkStatus === "idle" && (
        <p className={"animate-fade-up text-center text-2xl font-bold"}>
          {message}
        </p>
      )}
      {checkStatus === "checking" && (
        <p
          className={
            "animate-bounce text-center text-2xl font-bold animate-infinite animate-ease-in-out"
          }
        >
          Checking...
        </p>
      )}
      {checkStatus === "error" && (
        <p
          className={
            "animate-fade-up text-center text-2xl font-bold text-red-600"
          }
        >
          Something Wrong... :(
        </p>
      )}
    </div>
  );
};
