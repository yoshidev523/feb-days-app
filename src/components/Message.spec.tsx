import { CheckStatus } from "@/types/checkStatus";
import { render } from "@testing-library/react";
import { Message } from "@/components/Message";

test.each([["idle" as const], ["checking" as const], ["error" as const]])(
  "renders correctly %s",
  (checkStatus: CheckStatus) => {
    const { asFragment } = render(
      <Message checkStatus={checkStatus} message={"Message"} />,
    );
    expect(asFragment()).toMatchSnapshot();
  },
);
