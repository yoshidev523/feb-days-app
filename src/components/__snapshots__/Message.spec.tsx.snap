// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly checking 1`] = `
<DocumentFragment>
  <div
    data-testid="message"
  >
    <p
      class="animate-bounce text-center text-2xl font-bold animate-infinite animate-ease-in-out"
    >
      Checking...
    </p>
  </div>
</DocumentFragment>
`;

exports[`renders correctly error 1`] = `
<DocumentFragment>
  <div
    data-testid="message"
  >
    <p
      class="animate-fade-up text-center text-2xl font-bold text-red-600"
    >
      Something Wrong... :(
    </p>
  </div>
</DocumentFragment>
`;

exports[`renders correctly idle 1`] = `
<DocumentFragment>
  <div
    data-testid="message"
  >
    <p
      class="animate-fade-up text-center text-2xl font-bold"
    >
      Message
    </p>
  </div>
</DocumentFragment>
`;
